external/fv3fit/fv3fit/reservoir/adapters.py

```diff
@@ -0,0 +1,169 @@
+        if len(arr.shape) == 3 and arr.shape[-1] > 1:
+            dims.append("z")
+        return xr.DataArray(data=arr, dims=dims)
+
+    def output_array_to_ds(
+        self, outputs: Sequence[np.ndarray], output_dims: Sequence[str]
+    ) -> xr.Dataset:
+        return xr.Dataset(
+            {
+                var: self._ndarray_to_dataarray(output)
+                for var, output in zip(self.output_variables, outputs)
+            }
+        ).transpose(*output_dims)
+
+    def input_dataset_to_arrays(
+        self, inputs: xr.Dataset, variables: Iterable[Hashable]
+    ) -> Sequence[np.ndarray]:
+        # Converts from xr dataset to sequence of variable ndarrays expected by encoder
+        # Make sure the xy dimensions match the rank divider
+        transposed_inputs = _transpose_xy_dims(ds=inputs, rank_dims=self.rank_dims)
+        input_arrs = []
+        for variable in variables:
+            da = transposed_inputs[variable]
+            if "z" not in da.dims:
+                da = da.expand_dims("z", axis=-1)
+            input_arrs.append(da.values)
+        return input_arrs
+
+
+@io.register("reservoir-adapter")
+class ReservoirDatasetAdapter(Predictor):
+    MODEL_DIR = "reservoir_model"
+
+    def __init__(
+        self,
+        model: ReservoirComputingModel,
+        input_variables: Iterable[Hashable],
+        output_variables: Iterable[Hashable],
+    ) -> None:
+        self.model = model
+        self.input_variables = input_variables
+        self.output_variables = output_variables
+        self.model_adapter = DatasetAdapter(
+            input_variables=input_variables,
+            output_variables=output_variables,
+            rank_dims=model.rank_divider.rank_dims,
+        )
+
+    def predict(self, inputs: xr.Dataset) -> xr.Dataset:
+        # inputs arg is not used, but is required by Predictor signature and prog run
+        prediction_arr = self.model.predict()
+        return self.model_adapter.output_array_to_ds(
+            prediction_arr, output_dims=list(inputs.dims)
+        )
+
+    def increment_state(self, inputs: xr.Dataset):
+        xy_input_arrs = self.model_adapter.input_dataset_to_arrays(
+            inputs, self.input_variables
+        )  # x, y, feature dims
+        self.model.increment_state(xy_input_arrs)
+
+    def reset_state(self):
+        self.model.reset_state()
+
+    def dump(self, path):
+        self.model.dump(os.path.join(path, self.MODEL_DIR))
+
+    @classmethod
+    def load(cls, path: str) -> "ReservoirDatasetAdapter":
+        model = ReservoirComputingModel.load(os.path.join(path, cls.MODEL_DIR))
+        model.reset_state()
+        adapter = cls(
+            input_variables=model.input_variables,
+            output_variables=model.output_variables,
+            model=model,
+        )
+        return adapter
+
+
+@io.register("hybrid-reservoir-adapter")
+class HybridReservoirDatasetAdapter(Predictor):
+    MODEL_DIR = "hybrid_reservoir_model"
```

Comment:
Perhaps add some light docstrings explaining the usage case of each adapter and when arguments are used (or not used).  E.g., the hybrid input variables used in the predict while inputs are used in the increment only.  Really only need the model to use in the adapter, but I/O variables are required for the dump/load infra.